### PR TITLE
feat(external-fallbacks): enable modules to have external fallbacks

### PR DESCRIPTION
**Pull request by JAdshead**
This PR requires release and update of Holocron with the changes outlined in https://github.com/americanexpress/holocron/pull/143

## Description

Enable modules which require externals to use a fallback when the root module does not provide one.
Record each modules required externals when the module is required by one-app.
When rendering html to the client include all the fallback script tags required for the modules used in the initial render.
pass the registry of required externals down to the client then hydrate when client initializes 

## Motivation and Context

This will allow more flexibility to holocron modules when a required external is missing. It also provides an additional upgrade path for major versions of external dependencies. 

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (adding or updating documentation)
- [ ] Dependency update
- [ ] Security update

## Checklist:

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] These changes should be applied to a maintenance branch.
- [ ] This change requires cross browser checks.
- [ ] Performance tests should be ran against the server prior to merging.
- [ ] This change impacts caching for client browsers.
- [ ] This change impacts HTTP headers.
- [ ] This change adds additional environment variable requirements for One App users.
- [ ] I have added the Apache 2.0 license header to any new files created.

### CODE REVIEW

File: src/server/utils/onModuleLoad.js [hunk 1]
```diff
@@ -32,25 +33,12 @@ import { setErrorPage } from '../plugins/reactHtml/staticErrorPage';
...
-
-export const getModulesUsingExternals = () => modulesUsingExternals.toJS();
 
-export const setModulesUsingExternals = (moduleNames) => {
-  modulesUsingExternals = new ImmutableSet(moduleNames);
-};
+export const CONFIGURATION_KEY = 'appConfig';
 
-const logModuleLoad = (moduleName, moduleVersion) => {
+function logModuleLoad(moduleName, moduleVersion) {
```
Comment (Matthew-Mallimo) [inline src/server/utils/onModuleLoad.js:39]: why was this changed?

File: src/server/utils/onModuleLoad.js [hunk 2]
```diff
@@ -147,32 +133,6 @@ export default function onModuleLoad({
     );
   }
 
-  if (requiredExternals) {
```
Comment (Matthew-Mallimo) [inline src/server/utils/onModuleLoad.js:150]: Did all this logic get moved to holocron?

File: package.json [hunk 3]
```diff
@@ -108,7 +108,7 @@
     "fastify": "^4.10.2",
     "fastify-plugin": "^4.2.0",
     "helmet": "^6.0.0",
-    "holocron": "^1.7.0",
+    "holocron": "file:holocron-v1.7.0.tgz",
```
Comment (Matthew-Mallimo) [inline package.json:111]: This will need to be fixed before merging

**Review by Matthew-Mallimo — CHANGES_REQUESTED** [review]

File: src/client/initClient.jsx [hunk 4]
```diff
@@ -29,6 +29,8 @@ export default async function initClient() {
   try {
     // eslint-disable-next-line no-underscore-dangle
     setModuleMap(global.__CLIENT_HOLOCRON_MODULE_MAP__);
+    // eslint-disable-next-line no-underscore-dangle
+    setRequiredExternalsRegistry(global.__holocron_externals__);
```
[the remark this answers is not in the document]
Comment (giulianok) [reply src/client/initClient.jsx:33]: good point, there are still other lower case global vars but I like the consistency of using all caps for it

File: src/server/plugins/reactHtml/index.jsx [hunk 5]
```diff
@@ -87,12 +99,45 @@ export function renderModuleScripts({
...
+      const { baseUrl } = modules[moduleName];
+      const endsWithSlash = baseUrl.endsWith('/');
+      const src = `${baseUrl}${endsWithSlash ? '' : '/'}${name}.browser.js`;
+      return renderScript({
+        src,
+        integrity,
+        isDevelopmentEnv,
+        clientCacheRevision,
+      });
+    }).join(isDevelopmentEnv ? '\n          ' : '');
```
[the remark this answers is not in the document]
Comment (giulianok) [reply src/server/plugins/reactHtml/index.jsx:138]: idk to be honest. @JAdshead any reason why you've added these spaces?

File: src/server/plugins/reactHtml/index.jsx [hunk 6]
```diff
@@ -87,12 +99,45 @@ export function renderModuleScripts({
...
+    return renderScript({
+      src, integrity, isDevelopmentEnv, clientCacheRevision,
+    });
   }).join(isDevelopmentEnv ? '\n          ' : '');
 }
 
+function isDuplicateExternal({ name, version }, currentExternals) {
+  return currentExternals
+    .some((current) => (name === current.name) && (version === current.version));
+}
```
[the remark this answers is not in the document]
Comment (giulianok) [reply src/server/plugins/reactHtml/index.jsx:111]: if you are referring to the `filter` + `isDuplicateExternal` then, technically, it's `O(nm)`, and yes, can be `O(n^2)` depending on the data it each array holds. Also, both arrays are small, no big impact, but I'm all for performance so happy to convert it into a set for quick lookup

File: src/server/utils/onModuleLoad.js (hunk 1 above)
Comment (giulianok) [reply src/server/utils/onModuleLoad.js:39]: consistency

File: src/server/utils/onModuleLoad.js (hunk 2 above)
Comment (giulianok) [reply src/server/utils/onModuleLoad.js:150]: yes

File: __tests__/server/utils/getModulesToUpdate.spec.js [hunk 7]
```diff
@@ -152,7 +152,7 @@ describe('getModulesToUpdate', () => {
         },
       },
     };
-    expect(getModulesToUpdate(current, next)).toEqual(['first-module', 'another-module', 'test-root']);
+    expect(getModulesToUpdate(current, next)).toEqual(['test-root']);
```
Comment (Matthew-Mallimo) [inline __tests__/server/utils/getModulesToUpdate.spec.js:155]: Why did this expect change if the test case didn't change?

File: package.json [hunk 8]
```diff
@@ -31,7 +31,7 @@
     "test:lockfile": "lockfile-lint -p package-lock.json -t npm -a npm -o https: -c -i",
     "test:lint": "eslint --ext js,jsx,md,snap .",
     "start": "node lib/server/index.js",
-    "start:inspect": "node --inspect --expose-gc lib/server/index.js",
+    "start:inspect": "node --inspect-brk --expose-gc lib/server/index.js",
```
Comment (Matthew-Mallimo) [inline package.json:34]:
```suggestion
    "start:inspect": "node --inspect --expose-gc lib/server/index.js",
```

**Review by Matthew-Mallimo — CHANGES_REQUESTED** [review]

File: __tests__/server/utils/getModulesToUpdate.spec.js (hunk 7 above)
Comment (JAdshead) [reply __tests__/server/utils/getModulesToUpdate.spec.js:155]: good catch, looks like the mock outdated

File: package.json (hunk 8 above)
Comment (JAdshead) [reply package.json]: resolved in 5072f48bf0080408e9584a2277514d411e3298ef

File: __tests__/server/utils/getModulesToUpdate.spec.js (hunk 7 above)
Comment (JAdshead) [reply __tests__/server/utils/getModulesToUpdate.spec.js]: resolved 5072f48bf0080408e9584a2277514d411e3298ef

**Review by Matthew-Mallimo — APPROVED** [review]

**Review by Matthew-Mallimo — APPROVED** [review]

**Review by Matthew-Mallimo — DISMISSED** [review]

File: src/client/initClient.jsx [hunk 9]
```diff
@@ -29,6 +29,8 @@ export default async function initClient() {
   try {
     // eslint-disable-next-line no-underscore-dangle
     setModuleMap(global.__CLIENT_HOLOCRON_MODULE_MAP__);
+    // eslint-disable-next-line no-underscore-dangle
+    setRequiredExternalsRegistry(global.__HOLOCRON_EXTERNALS__);
```
Comment (code-forger) [inline src/client/initClient.jsx:33]: NIT: for consistency with `__CLIENT_HOLOCRON_MODULE_MAP__`, why not `__CLIENT_HOLOCRON_EXTERNALS__`?

File: src/server/utils/onModuleLoad.js [hunk 10]
```diff
@@ -33,21 +33,8 @@ import { setRedirectAllowList } from './redirectAllowList';
...
-
-const clearModulesUsingExternals = () => {
-  modulesUsingExternals = modulesUsingExternals.clear();
-};
 
-export const getModulesUsingExternals = () => modulesUsingExternals.toJS();
-
-export const setModulesUsingExternals = (moduleNames) => {
-  modulesUsingExternals = new ImmutableSet(moduleNames);
-};
```
Comment (code-forger) [inline src/server/utils/onModuleLoad.js:50]: Removing parts of the old system? is this still backwards compatible and this functionality is now achieved in a different way?

**Review by code-forger — APPROVED** [review]
Approved with 1 NIT and 1 question

File: src/server/utils/onModuleLoad.js (hunk 10 above)
Comment (JAdshead) [reply src/server/utils/onModuleLoad.js:50]: for the most part this has been reimplemented in holocron - https://github.com/americanexpress/holocron/pull/143/files#diff-c59d732ae8137c4f28a025d9cabd09b4fd3166326f1822eb6b9a94089d89d777R68-R77

it does appear that the call to `clearModulesUsingExternals` has been removed when it should not have.

File: src/client/initClient.jsx (hunk 9 above)
Comment (JAdshead) [reply src/client/initClient.jsx:33]: `__CLIENT_HOLOCRON_MODULE_MAP__` is the only key with the `CLIENT` prefix, this is probably as the module map being set is specifically for the client and differs from the module map loaded on the server.

**Review by Matthew-Mallimo — APPROVED** [review]

File: src/server/plugins/reactHtml/index.jsx [hunk 11]
```diff
@@ -88,12 +100,60 @@ export function renderModuleScripts({
...
 }
 
+export function renderExternalFallbacks({
+  clientInitialState,
+  moduleMap,
+  isDevelopmentEnv,
+  isLegacy,
+}) {
+  const loadedModules = clientInitialState.getIn(['holocron', 'loaded'], iSet()).toArray();
+  const requiredFallbacks = loadedModules
```
Comment (dogpatch626) [inline src/server/plugins/reactHtml/index.jsx:116]: this is a pretty expensive looking reduce in terms of time. are we bound to this by the shape of the data that is in loadedModules or is there potentially a better way of doing this. nested objects with arrays are fun 😄

Comment (JAdshead) [reply src/server/plugins/reactHtml/index.jsx:116]: This would only be a handful of modules. loadedModules are modules which are required to render the page, not all the modules included in the module map. even them i'm not sure a module map would be ever be large enough to provide performance concerns. 

> are we bound to this by the shape of the data that is in loadedModules
yes, this could be rewritten to remove that additional reduce on 118 but im not sure that it would add much value.

**Review by dogpatch626 — APPROVED** [review]

**Review by Matthew-Mallimo — APPROVED** [review]
